subprojects/core/src/main/java/org/gradle/api/internal/file/DefaultFileOperations.java

```diff
@@ -178,37 +178,30 @@ public FileTreeInternal zipTree(Object zipPath) {
 
     @Override
     public FileTreeInternal tarTree(Object tarPath) {
+        if (tarPath instanceof ReadableResource) {
+            boolean hasBackingFile = tarPath instanceof ReadableResourceInternal
+                && ((ReadableResourceInternal) tarPath).getBackingFile() != null;
+            if (!hasBackingFile) {
+                throw new InvalidUserCodeException("Cannot use tarTree() on a resource without a backing file.");
+            }
+        }
```

Comment:
I think we should move this to `asFileProvider()` instead.

Reply:
Done. This allowed me to simplify `TarFileTree` a little bit more as well.